### ISSUE TITLE
DIST: Improve method of distributing packages to users

### ISSUE THREAD

**Issue by asoplata** [jonescompneurolab/hnn-core#950]
### Problem:
The current install method suffices for non-parallel simulation, but in an ideal world, it would be easy for users to install both HNN AND what it needs to use the incredible speed-up offered by MPI. Let's make that happen: I propose we change the *default* install method and instructions to include MPI support (that's the install *method*, which is not the same thing as `pip install hnn_core`).

Cross-platform distribution of MPI binaries (and libraries) can be difficult, but Anaconda (Conda) rectifies this by itself being an MPI binary distribution platform; see https://anaconda.org/conda-forge/openmpi and https://anaconda.org/conda-forge/mpi4py .

### Solution:
Therefore, I think the best way to do this is to change our default install method by **distributing HNN-Core as its own Conda package**. This will allow us to specify the right version of Conda's OpenMPI packages as dependencies, which is not possible using the Python ecosystem alone (such as using wheels).

### Required Work:
This is not as straightforward as it sounds. We need to first investigate the dependency network between the primary Conda package we need https://anaconda.org/conda-forge/mpi4py (which depends on Conda-installed OpenMPI binaries) across platforms and against NEURON itself. For example, even though conda-forge ships a version of NEURON ( https://anaconda.org/conda-forge/neuron/files ), it is not guaranteed that the NEURON distributables are built against the same OpenMPI version that a conda-forge `mpi4py` is. Similarly, conda-forge NEURON only has distributables for x86_64 Linux and Mac, but not Apple Silicon (arm64) or Windows. Our Conda package may require installing the conda-forge version of `mpi4py`, but also depend on the `pip` version of NEURON. Similarly, we may not be able to improve the Windows distribution using this method; it may still be required that the user manually install NEURON, then install HNN-Core.

### Notes:
Please note that this is a task about *what form we use for distributing our code/assets to users (the “user side”)*. This is distinct from improvements to *how do we build our packages using development tools (the “dev side”).*, such as here #951 .

**Comment by jasmainak** [comment]
I thought the issue with `mpi4py` is that you need to have either OpenMPI or another MPI binary installed on your computer ... hence the need to go through anaconda. The anaconda install gets you both the MPI binary and `mpi4py`. Can this be done through wheels?

**Comment by asoplata** [comment]
@jasmainak No, I don't think that distributing wheels allows for a reliable method of simultaneously distributing MPI binaries. I think you're correct: cross-platform MPI binary distribution itself is a difficult situation, but fortunately for us, `Anaconda` is *arguably* the current-best way to handle that distribution, e.g. https://anaconda.org/conda-forge/openmpi . I am going to re-write this Issue description for how we should investigate building a better install method that includes MPI by default for our users.

**Comment by asoplata** [comment]
Coming back to this: in tandem with our comprehensive websites-overhaul, I think that now may be a good time to start working on this, so that it is ready for our April Workshop.

**Comment by asoplata** [comment]
@ntolley @dylansdaniels  We need to discuss what we want our future `conda` package installer ways to look like.

Firstly, unfortunately conda currently has no equivalent to `pip`'s "`extras`" https://github.com/jonescompneurolab/hnn-core/blob/master/setup.py#L78-L86 which we strongly rely on for providing variant installers. It has been requested for many years by many people, and while there is activity towards it right now, I have no idea how long it will take (months? years?). If you are interested, here are some links (putting then in a code block in order to not "mention" them):

```
main spec work, active https://github.com/conda/ceps/pull/111
https://github.com/conda/conda/issues/7502
https://github.com/conda/conda/issues/11053
https://stackoverflow.com/questions/42587385/install-extras-with-conda
```

In the meantime, as some of the discussions like in 7502 above have pointed out, a common workaround is to simply publish multiple packages in conda, where there's a separate package for each variant. This would imply that the equivalent of, for example, `pip install "hnn_core[gui,parallel]"` would be something like `conda install hnn-core-gui hnn-core-parallel` etc., which I think is reasonable. This is obviously a bigger maintenance burden than trying to provide a single conda package, but it follows the same philosophy as our current install landscape.

Here's a proposal for the list of separate `conda` packages to create:

- `hnn-core`: the minimal core, just like installing `pip install hnn-core`.
- `hnn-core-gui`: identical to `pip install "hnn-core[gui]"`, installs minimal `hnn-core` along with the `gui` dependencies.
- `hnn-core-opt`: identical to `pip install "hnn-core[opt]"`, installs minimal `hnn-core` along with `scikit-learn`.
- `hnn-core-parallel`: This is the entire reason for using conda in the first place. This will install the dependencies inherent to `pip install "hnn-core[parallel]"` in addition to the `conda-forge` versions of `mpi4py` and `openmpi>5` for MacOS and Linux. As usual, Windows support will be left for future work, if ever.
- `hnn-core-all`: (Optional) We could also provide a "meta-package" that simply installs all of the above. I'm 50/50 on this, but it is a common pattern used in package managers.
- Note: I don't see any reason for making an `hnn-core-docs` equivalent of `pip install "hnn-core[docs]"`.
- Note: There is no need for an `hnn-core-test` equivalent of  `pip install "hnn-core[test]"` for two reasons:
    1. There is already a way for advanced users to install the testable dependencies.
    2. All of the above conda packages will already be testable in-place, due to conda packages' support for testing ( https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#test-requirements ).

That covers all of the current install variants. 

All of the above is entirely up for discussion, of course.

**Comment by asoplata** [comment]
Initial testing of conda package building is going very well and ahead of schedule: I am able to build, install, and run (with all features) conda packages for MacOS arm64 and x86_64, and Linux x86_64. Currently, I'm doing testing using the scripts + recipe files located at https://github.com/asoplata/gh-testbench/tree/main/conda-packaging though note that that repo is known to change rapidly (it's only for random testing) and break its directory structure. There's still polishing left to do for the build-recipes (removing unnecessary dependencies, preparing them for importation to the conda-forge ecosystem, etc.).

I mention the above platforms specifically since:
1. Linux on arm64, more commonly known as `aarch64`: I don't think we should support explicit `conda` builds for this at this time, because: 
    1. AFAIK, no HNN developers have access to any computers with that hardware, for testing
    2. As far as I can tell, Linux on Arm is only common for server usage, and very rare for personal computers. On top of that, any servers/HPCs that our users are using are probably for scientific use, and are likely using x86_64 processors instead of arm. *On top of that*, when we publish installation instructions that include installing our conda package, we need to make clear that the conda packages should NOT be used on HPC installs. Instead, users need to follow their HPC's documentation for how to handle MPI loading and probably work with us to get them setup.
    3. Anyone using Linux on Arm probably has the technical know-how to follow our non-conda, pip+MPI installation instructions.
    4. If a user needs this, then we can provide it for that user, and also use their help to help us test the build.

**Comment by asoplata** [comment]
@dylansdaniels @ntolley I also want to propose a new, simpler list of conda packages:
- `hnn-core`: as above, just the minimal install, with no extra features. This way, `conda install hnn-core` and `pip install hnn-core` are analogous (except for the special case of Windows! see above)
- `hnn-core-all`: HNN with all the dependencies for `gui`, `opt`, and `parallel` included (but not `docs` or `testing`). This will become the new, recommended local install for Mac and Linux, and will make the initial install instructions *considerably* shorter and easier.

**Comment by asoplata** [comment]
In the dev meeting on 2025-04-14, we agreed that we will approach doing the two-package approach.

**Comment by asoplata** [comment]
Note: I realized there is a case where we may soon see a few users that require a Linux on ARM build: if they install HNN on WSL, but on a Windows machine using an ARM processor. There are increasingly more consumer products using Windows on ARM64, though it's still going to be a small fraction of Windows users for probably many years. I don't think we need to provide that platform build at this time, unless a user gets in touch and can help us use it. *Eventually*, if we ever get our conda-package-building to build for different platforms using CI runners instead of local computers, it should be easy to add a Linux-ARM build.

**Comment by asoplata** [comment]
The two conda builds are available from https://anaconda.org/jonescompneurolab/repo . Documentation and scripts on *how* to build the packages are now available in a new repo at  https://github.com/jonescompneurolab/hnn-core-conda-packaging . Final step is to overhaul our Install page to suggest using `hnn-core-all` as the default user's first port-of-entry.
